stapdyn/stapdyn.8
.\" -*- nroff -*-
.TH STAPDYN 8 
.SH NAME
stapdyn \- systemtap dyninst runtime

.SH SYNOPSIS

.br
.B stapdyn
[
.I OPTIONS
]
.I MODULE
[
.I MODULE-OPTIONS
]

.SH DESCRIPTION

The
.I stapdyn
program is the dyninst back-end of the Systemtap tool.  It expects a 
shared library produced by the front-end
.I stap
tool, when run with
.IR \-\-dyninst .

.PP
Splitting the systemtap tool into a front-end and a back-end allows a
user to compile a systemtap script on a development machine that has
the debugging information (need to compile the script) and then
transfer the resulting shared objevct to a production machine that
doesn't have any development tools or debugging information installed.
.PP
Please refer to stappaths (7) for the version number, or run
rpm \-q systemtap (fedora/red hat)
apt\-get \-v systemtap (ubuntu)

.SH OPTIONS
The
.I stapdyn
program supports the following options.  Any other option
prints a list of supported options.
.TP
.B \-v
Verbose mode.
.TP
.B \-V
Print version number and exit.
.TP
.B \-w
Suppress warnings from the script.
.TP
.B \-c CMD
Command CMD will be run and the
.I stapdyn
program will exit when CMD
does.  The '_stp_target' variable will contain the pid for CMD.
.TP
.B \-x PID
The '_stp_target' variable will be set to PID.
.TP
.B \-o FILE
Send output to FILE. If the module uses bulk mode, the output will
be in percpu files FILE_x(FILE_cpux in background and bulk mode)
where 'x' is the cpu number. This supports strftime(3) formats
for FILE.
.TP
.B \-C WHEN
Control coloring of error messages. WHEN must be either
.nh
"never", "always", or "auto"
.hy
(i.e. enable only if at a terminal). If the option is missing, then "auto"
is assumed. Colors can be modified using the SYSTEMTAP_COLORS environment
variable. See the
.IR stap (1)
manual page for more information on syntax and behaviour.
.TP
.B var1=val
Sets the value of global variable var1 to val. Global variables contained 
within a script are treated as options and can be set from the 
stapdyn command line.

.SH ARGUMENTS
.B MODULE
is either a module path or a module name.  If it is a module name,
the module will be looked for in the following directory
(where 'VERSION' is the output of "uname \-r"):
.IP
/lib/modules/VERSION/systemtap
.PP

\& $ stap \-\-dyninst \-p4 \-m mod1 \-e\ \[aq]global var1="foo"; probe begin{printf("%s\\n", var1); exit()}\[aq]
.br
.PP
Running this with an additional module argument:
.PP

\& $ stapdyn mod1.so var1="HelloWorld"
.br
\& HelloWorld
.PP
Spaces and exclamation marks currently cannot be passed into global variables 
this way.

.SH EXAMPLES
See the 
.IR stapex (3stap)
manual page for a collection of sample scripts.
.PP
Here is a very basic example of how to use
.I stapdyn.
First, use
.I stap
to compile a script.  The
.I stap
program will report the pathname to the resulting module.
.PP
\& $ stap \-\-dyninst \-p4 \-e \[aq]probe begin { printf("Hello World!\\n"); exit() }\[aq]
.br
\& /home/user/.systemtap/cache/85/stap_8553d83f78c_265.so
.PP
Run
.I stapdyn
with the pathname to the module as an argument.
.PP
\& $ stapdyn /home/user/.systemtap/cache/85/stap_8553d83f78c_265.so
.br
\& Hello World!

.SH SAFETY AND SECURITY
Systemtap, in DynInst mode, is a developer tool, and runs completely
unprivileged.  The Linux kernel will only permit one's own processes
to be accessed, which is enforced by the
.IR ptrace (2)
system call.
See the 
.IR stap (1)
manual page for additional information on safety and security.

.SH SEE ALSO
.IR stap (1),
.IR stapprobes (3stap),
.IR stap\-server (8),
.IR staprun (8),
.IR stapex (3stap)

.SH BUGS
Use the Bugzilla link of the project web page or our mailing list.
.nh
.BR http://sourceware.org/systemtap/ ", " <systemtap@sourceware.org> .
.hy



NEWS
* What's new in version 2.3

- java("org.my.MyApp") probes are now restricted to pre-existing jvm pid's with
  a listing in jps -l output to avoid recursive calls

- The tapset [nd_]syscall.semop parameter tsops_uaddr is renamed sops_uaddr for
  consistency with [nd_]syscall.semtimedop.

- The udp.stp tapset adds some ip-address/port variables.

- A new guru-mode-only tapset function raise() is available to send signals
  to the current task.

- The following tapset functions are deprecated in release 2.3:
  'stap_NFS_CLIENT', '__getfh_inode', '_success_check',
  '_sock_prot_num', '_sock_fam_num', '_sock_state_num',
  '_sock_type_num', and '_sock_flags_num'.

* What's new in version 2.2.1, 2013-05-16
* What's new in version 2.2, 2013-05-14

- Experimental support has been added for probing Java methods using
  Byteman 2.0 as a backend. Java method probes can target method entries,
  returns, or specific statements in the method as specified by line number.

      probe java("org.my.MyApp").class("^java.lang.Object").method("foo(int)")
         { println($$parms) }

  See java/README for information on how to set up Java/Byteman
  functionality.  Set env STAPBM_VERBOSE=yes for more tracing. 

- The stap -l output and pn() tapset function's return value may be slightly
  different for complicated web of wildcarded/aliased probes.

- The dyninst backend has improved in several aspects:

  - Setting custom values for global variables is now supported, both
    with -G when compiling a script, and from the stapdyn command line
    when loading a precompiled module.

  - A high-performance shared-memory-based transport is used for
    trace data.

- A systemd service file and tmpfile have been added to allow
  systemtap-server to be managed natively by systemd.

- Due to the removal of register_timer_hook in recent kernels, the
  behaviour of timer.profile has been changed slightly. This probe is
  now an alias which uses the old mechanism where possible, but falls
  back to perf.sw.cpu_clock or another mechanism when the kernel timer
  hook is not available.

  To require the kernel timer hook mechanism in your script, use
  timer.profile.tick instead of timer.profile.

- The following tapset variables are deprecated in release 2.2:
  - The 'origin' variables in the 'generic.fop.llseek',
    'generic.fop.llseek.return', and 'nfs.fop.llseek' probes. The
    'origin' variable has been replaced by the 'whence' variable.
  - The 'page_index' variable in the 'vfs.block_sync_page' and
    'vfs.buffer_migrate_page' probe aliases.
  - The 'write_from' and 'write_upto' variables in the
    '_vfs.block_prepare_write' and '_vfs.block_prepare_write.return'
    probe aliases.
  - The 'regs' variable in the 'syscall.sigaltstack',
    'nd_syscall.sigaltstack', 'syscall.fork', and 'nd_syscall.fork'
    probe aliases.
  - The 'first', 'second', 'third', and 'uptr_uaddr' variables in the
    'syscall.compat_sys_shmat' and 'nd_syscall.compat_sys_shmat' probe
    aliases.

- The following tapset functions are deprecated in release 2.2:
  'ppos_pos', '_dev_minor', and '_dev_major'

- The folowing tapset functions used to return error strings instead
  of raising an error. The original behavior is deprecated in release
  2.2.

  'ctime', 'probemod', 'modname'

* What's new in version 2.1, 2013-02-13

- EMACS and VIM editor modes for systemtap source files are included / updated.

- The translator now eliminates duplicate tapset files between its
  preferred directory (as configured during the build with --prefix=/
  or specified with the -I /path option), and files it may find under
  $XDG_DATA_DIRS.  This should eliminate a class of conflicts between
  parallel system- and hand-built systemtap installations.

- The translator accepts a --suppress-time-limits option, which defeats
  time-related constraints, to allows probe handlers to run for indefinite
  periods.  It requires the guru mode (-g) flag to work.  Add the earlier
  --suppress-handler-errors flag for a gung-ho "just-keep-going" attitude.

- Perf event probes may now be read on demand. The counter probe is
  defined using the counter-name part: 
  probe perf.type(0).config(0).counter("NAME").  The counter is
  read in a user space probe using @perf("NAME"), e.g.
   process("PROCESS").statement("func@file") {stat <<< @perf("NAME")} 

- Perf event probes may now be bound to a specific task using the
  process-name part:  probe perf.type(0).config(0).process("NAME") { }
  If the probed process name is not specified, then it is inferred
  from the -c CMD argument.  

- Some error messages and warnings now refer to additional information
  that is found in man pages.  These are generally named
  error::FOO or warning::BAR (in the 7stap man page section)
  and may be read via
     % man error::FOO
     % man warning::BAR

- The dyninst backend has improved in several aspects:
  - The runtime now allows much more concurrency when probing multithreaded
    processes, and will also follow probes across forks.
  - Several new probe types are now supported, including timers, function
    return, and process.begin/end and process.thread.begin/end.
  - Semaphores for SDT probes are now set properly.
  - Attaching to existing processes with -x PID now works.

- The foreach looping construct can now sort aggregate arrays by the user's
  choice of aggregating function.  Previously, @count was implied.  e.g.:
     foreach ([x,y] in array @sum +) { println(@sum(array[x,y])) } 

- Proof of concept support for regular expression matching has been added:
     if ("aqqqqqb" =~ "q*b") { ... }
     if ("abc" !~ "q*b") { ... }

  The eventual aim is to support roughly the same functionality as
  the POSIX Extended Regular Expressions implemented by glibc.
  Currently missing features include extraction of the matched string
  and subexpressions, and named character classes ([:alpha:], [:digit:], &c).

  Special thanks go to the re2c project, whose public domain code this

stapdyn/stapdyn.cxx
// stapdyn main program
// Copyright (C) 2012-2013 Red Hat Inc.
//
// This file is part of systemtap, and is free software.  You can
// redistribute it and/or modify it under the terms of the GNU General
// Public License (GPL); either version 2, or (at your option) any
// later version.

#include <iostream>
#include <memory>

extern "C" {
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <wordexp.h>
}

#include "config.h"
#include "../git_version.h"
#include "../version.h"

#include "mutator.h"
#include "dynutil.h"


#define _STRINGIFY(s) #s
#define STRINGIFY(s) _STRINGIFY(s)
#define DYNINST_FULL_VERSION \
  ( STRINGIFY(DYNINST_MAJOR) "." \
    STRINGIFY(DYNINST_MINOR) "." \
    STRINGIFY(DYNINST_SUBMINOR) )


using namespace std;


// Print the obligatory usage message.
static void __attribute__ ((noreturn))
usage (int rc)
{
  clog << "Usage: " << program_invocation_short_name
       << " MODULE [-v] [-c CMD | -x PID] [-o FILE] [-C WHEN] [globalname=value ...]" << endl
       << "-v              Increase verbosity." << endl
       << "-c cmd          Command \'cmd\' will be run and " << program_invocation_short_name << " will" << endl
       << "                exit when it does.  The '_stp_target' variable" << endl
       << "                will contain the pid for the command." << endl
       << "-x pid          Sets the '_stp_target' variable to pid." << endl
       << "-o FILE         Send output to FILE. This supports strftime(3)" << endl
       << "                formats for FILE." << endl
       << "-C WHEN         Enable colored errors. WHEN must be either 'auto'," << endl
       << "                'never', or 'always'. Set to 'auto' by default." << endl;

  exit (rc);
}


// This is main, the heart and soul of stapdyn, oh yeah!
int
main(int argc, char * const argv[])
{
  pid_t pid = 0;
  const char* command = NULL;
  const char* module = NULL;

  // Check if error/warning msgs should be colored
  color_errors = isatty(STDERR_FILENO)
    && strcmp(getenv("TERM") ?: "notdumb", "dumb");

  // First, option parsing.
  int opt;
  while ((opt = getopt (argc, argv, "c:x:vwo:VC:")) != -1)
    {
      switch (opt)
        {
        case 'c':
          command = optarg;
          break;

        case 'x':
          pid = atoi(optarg);
          break;

        case 'v':
          ++stapdyn_log_level;
          break;

        case 'w':
          stapdyn_suppress_warnings = true;
          break;

	case 'o':
	  stapdyn_outfile_name = optarg;
	  break;

        case 'V':
          fprintf(stderr, "Systemtap Dyninst loader/runner (version %s/%s, %s)\n"
                          "Copyright (C) 2012-2013 Red Hat, Inc. and others\n"
                          "This is free software; see the source for copying conditions.\n",
                  VERSION, DYNINST_FULL_VERSION, STAP_EXTENDED_VERSION);
          return 0;

        case 'C':
          if (!strcmp(optarg, "never"))
            color_errors = false;
          else if (!strcmp(optarg, "auto"))
            color_errors = isatty(STDERR_FILENO)
              && strcmp(getenv("TERM") ?: "notdumb", "dumb");
          else if (!strcmp(optarg, "always"))
            color_errors = true;
          else {
            staperror() << "Invalid option '" << optarg << "' for -C." << endl;
            usage (1);
          }
          break;
        default:
          usage (1);
        }
    }

  // The first non-option is our stap module, required.
  if (optind < argc)
    module = argv[optind++];

  // Remaining non-options, if any, specify global variables.
  vector<string> modoptions;
  while (optind < argc)
    {
      modoptions.push_back(string(argv[optind++]));
    }

  if (!module || (command && pid))
    usage (1);

  // Make sure that environment variables and selinux are set ok.
  if (!check_dyninst_rt())
    return 1;
  if (!check_dyninst_sebools(pid != 0))
    return 1;

  auto_ptr<mutator> session(new mutator(module, modoptions));
  if (!session.get() || !session->load())
    {
      staperror() << "Failed to create the mutator!" << endl;
      return 1;
    }

  if (command && !session->create_process(command))
    return 1;

  if (pid && !session->attach_process(pid))
    return 1;

  if (!session->run())
    return 1;

  return session->exit_status();
}

/* vim: set sw=2 ts=8 cino=>4,n-2,{2,^-2,t0,(0,u0,w1,M1 : */
